gapii/cc/pack_encoder.cpp

```diff
@@ -30,70 +31,77 @@ using ::google::protobuf::Message;
-const uint64_t NO_ID               = 0xffffffffffffffff;
+const int      VERSION_MAJOR          = 1;
+const int      VERSION_MINOR          = 1;
+const int64_t  TAG_FIRST_GROUP        = -1;
+const int64_t  TAG_GROUP_FINALIZER    = 0;
+const int64_t  TAG_DECLARE_TYPE       = 1;
+const int64_t  TAG_FIRST_OBJECT       = 2;
+const uint64_t NO_ID                  = 0xffffffffffffffff;
+
+constexpr int TYPE_ID_CACHE_COUNT  = 4;
 
 const char magic[] = "protopack";
 
 // PackEncoderImpl implements the PackEncoder interface.
 class PackEncoderImpl : public gapii::PackEncoder {
 public:
-    PackEncoderImpl(const std::shared_ptr<core::StreamWriter>& output);
+    PackEncoderImpl(const std::shared_ptr<core::StringWriter>& writer);
     ~PackEncoderImpl();
 
     virtual void object(const Message* msg) override;
     virtual SPtr group(const Message* msg) override;
 
 private:
+    struct TypeIDCache {
+        std::mutex mutex;
+        std::unordered_map<const Descriptor*, uint32_t> type_ids;
+    };
+
     struct Shared {
-        Shared(const std::shared_ptr<core::StreamWriter>& writer);
+        Shared(const std::shared_ptr<core::StringWriter>& writer);
 
         std::mutex mutex;
-        std::shared_ptr<core::StreamWriter> writer;
+        std::shared_ptr<core::StringWriter> writer;
         std::unordered_map<const Descriptor*, uint32_t> type_ids;
+        TypeIDCache type_id_caches[TYPE_ID_CACHE_COUNT];
         uint64_t group_id;
     };
 
     PackEncoderImpl(const std::shared_ptr<Shared>& shared, uint64_t group_id);
 
-    void writeGroupID();
+    void writeGroupID(std::string& buffer);
     uint64_t writeTypeIfNew(const Descriptor* desc);
-    void writeString(const std::string& str);
-    void writeVarint32(uint32_t value);
-    void writeZigzag(int64_t value);
-    void writeVarint(uint64_t value);
-    void writeVarintDirect(uint64_t value);
-    void flushChunk();
+    uint64_t writeTypeIfNewBlocking(const Descriptor* desc);
+    void writeString(std::string& buffer, const std::string& str);
+    void writeVarint32(std::string& buffer, uint32_t value);
+    void writeZigzag(std::string& buffer, int64_t value);
+    void writeVarint(std::string& buffer, uint64_t value);
+    void flushBuffer(std::string& buffer);
 
     std::shared_ptr<Shared> mShared;
     uint64_t mGroupId;
-
-    std::string mBuffer; // Flushes to mShared->mWriter
 };
 
-PackEncoderImpl::Shared::Shared(const std::shared_ptr<core::StreamWriter>& writer)
-        : group_id(0)
-        , writer(writer) {}
+PackEncoderImpl::Shared::Shared(const std::shared_ptr<core::StringWriter>& writer)
+        : writer(writer)
+        , group_id(0) {}
 
-PackEncoderImpl::PackEncoderImpl(const std::shared_ptr<core::StreamWriter>& writer)
+PackEncoderImpl::PackEncoderImpl(const std::shared_ptr<core::StringWriter>& writer)
         : mShared(new Shared(writer))
         , mGroupId(NO_ID) {
-    writer->write(magic, sizeof(magic) - 1);
+    auto stream = writer->getStream();
```

Comment:
You could kill the rather smelly `StringWriter::getStream()` method if you changed this constructor to take the `StreamWriter`, and built the `PackStringWriter::create(stream)` in here.